615 manage text data sets with muliple line text fields

marcboulle opened the pull request.

marcboulle committed:
Improve management of csv field encoding involving double-quotes

Rappel: Khiops ne gere pas les champs multi-lignes pour les raisons suivantes:
- gestion des erreurs
  - un champ multi-lignes est specifie en utilisant des double-quote `"` en debut et fin de champs
  - en cas d'erreur de specification (double-quote final manquant), il n'y a pas de reprise sur erreur possible
    - avec python, une seule erreur de ce type provoque un crash indescriptible
    - il faut alors lire tout jusqu'à la fin de fichier
    - non robustesse: une seule erreur dans un fichier de tres grande taille invalide tout un traitement
- traitement des index de lignes
  -  avec des champs multi-ligne, on n'a plus l'equivalence entre ligne et enregistrement identifie par un index de ligne:
     tout devient alors plus complexe à gerer et à manipuler
     (une erreur sur un enregistrement renvoie à son index de ligne...)
- limites en volumetrie
- il n'est pas envisageable de faire evoluer Khiops sur ce point, mais il faudra que ceci soit bien documente

Tout champ multi-ligne doit alors etre prealablement recode, par exemple en remplacant les sauts de lignes internes aux champs multi-lignes
 (caracteres `\r` et `\n`) par des caracteres blancs).

Evolutions:
- tout champ commencant par un double-quote, mais ne terminant pas par un double quote,
  aboutit maintenant a ignorer la ligne toute entiere
- tout champ se terminant par un double-quote, mais ne commencant pas par un double quote,
  est desormais detecte avec un warning
- lors de l'ecriture, la norme d'encodage csv est desormais entierement restectee:
  tout champs contenant un double-quote ou un separateur de champs est desormais encode entre double quotes

Impacts principaux:
- KWDataTableDriverTextFile
  - Read: re-ecriture
    - on memet un message pour tout probleme de double quote
    - mais on ignore les record que dans le cas d'un double quote manquant en fin de champ
  - SkipMainRecord: impact simimaire
- InputBufferedFile
  - FieldMissingBeginDoubleQuote: nouveau type d'erreur
  - SkipField, SkipDoubleQuoteField: re-ecritue pour gerer correctement les problemes de double-quotes
  - GetPrevChar, GetPrevPrevChar: nouvelles methodes internes
- KWDataTableDriverSlice
  - ReadObject: erreur si probleme de double-quote, qui ne devrait pas se produire
- OutputBufferedFile
  - WriteField: ecriture systematique d'un champ entre double quote, s'il contient un delimiteur OU un double quote

Impacts LearningTest
- Recodage de certains jeux de donnees: 20newsgroups, Accidents...
- References
  - TestKhiops/Rules/BugMultiTableRegexRules
  - TestKhiops/Advanced/AnyChar
  - TestKhiops/Advanced/MixedAsciiUTF8Chars
  - TestKhiops/Advanced/ReadWriteDoubleQuotes
  - TestKhiops/Advanced/TestTabulationEverywhere
  - TestKhiops/Bugs/StarValue
  - TestKhiops/BugsMultiTables/BugRecette
  - TestKhiops/BugsMultiTables/BugSortQuote
  - TestKhiops/BugsMultiTables/BugSortQuoteBis
  - TestKhiops/NewV10/DetectFileFormatCrash2
  - TestKhiops/NewV10/DetectFileFormatMessy
  - TestKhiops/TextVariables/*

marcboulle committed:
Collect total encoding error number relative to missing double quotes

Imapacts principaux:
- KWDataTableDriver
  - Set|GetMissingDoubleQuoteEncodingErrorNumber: memorisation du nombre d'erreurs d'encodage
- KWDataTableDriverTextFile
  - Read, SkipMainRecord: incrementation du MissingDoubleQuoteEncodingErrorNumber
- KWDatabaseTask
  - GetMissingDoubleQuoteEncodingErrorTotalNumber: nombre total d'erreur d'encodage detectees
  - collecte du nombre total d'erreurs d'encodage dans les methodes de la tache

Impacts sur LearningTest, avec les taches ayant une variable partagee additionnelles
- TestKhiops/CrashTests/checkdatabase_KWDatabaseCheckTask
- TestKhiops/CrashTests/deploymodel_KWDatabaseTransferTask
- TestKhiops/CrashTests/trainclassifier_KDSelectionOperandSamplingTask
- TestKhiops/CrashTests/trainclassifier_KWClassifierEvaluationTask
- TestKhiops/CrashTests/trainclassifier_KWDatabaseBasicStatsTask
- TestKhiops/CrashTests/trainclassifier_KWDatabaseSlicerTask

marcboulle committed:
Generate a synthetic error message in case of encoding errors

Generation d'un message synthetique sur les erreur d'encodage dans Check database', 'Train model', 'Deploy model' et 'Evaluate model'.
Impacts principaux
- renomage general des methodes de type MissingDoubleQuoteEncodingErrorNumber en EncodingErrorNumber
- KWPredictorEvaluation
  - GetEncodingErrorNumber: memorisation du nombre d'erreur d'encodage
    - alimente par KWPredictorEvaluationTask::Evaluate
- KWDatabase
  - AddEncodingErrorMessage: centralisation de la forme du message d'erreur lie aux erreur d'encodage
  - appele dans
    - KWClassStats::KWClassStats
    - KWDatabaseCheckTask::CheckDatabase
    - KWDatabaseTransferTask::Transfer
    - KWPredictorEvaluator::EvaluatePredictors

Optimisation de l'affichage des erreurs detectees lors de l'analyse d'une base, pour eviter de les afficher plusieurs fois
- on se place en mode non verbeux dans les cas suivants
  - passe prealable de construction de variable, car la vase sera analysee ensuite pour la preparation
  - deuxieme passe de préparation en cas de repriuse sur erreur en cas de regression avec valeurs cibles manquantes
  - evaluation sur la base d'apprentissage, deja analysee
  - evaluation sur la base de test, si c'est la meme que la base d'apprentissage
- impact: KWLearningProblem::ComputeStats
- idem pour l'evaluation des predicteurs, ou on se place en mode non verbeux a partir
  du deuxieme predicteur evalue sur la meme base

Divers
  - supression de la classe MDKhiopsAnalysisSpec (code mort)

marcboulle committed:
Unify term 'double quote', rather than 'double-quote'

bruno-at-orange reviewed (approved):
LGTM
One comment about the error message.

marcboulle committed:
Extend generation of encoding errors to coclustering

La tache de coclustering lit directement la base en sequentiel, sans passer par une tache.
On a donc etendu la classe database pour memoriser les erreur d'encodage detectees
lors de la derniere passe de lecture

Imapcts:
- CCLearningProblemView::BuildCoclustering: generation du message d'erreur d'encodage
- KWDatabase
  - Set|GetEncodingErrorNumber
  - OpenForRead: initialisation a 0
- KWSTDatabase::PhysicalClose: memorisation des erreur d'encodage a la fermeture
- KWMTDatabase::DMTMPhysicalClose: memorisation des erreur d'encodage a la fermeture pour chaque table du schema

marcboulle committed:
Refactor management of encoding errors within KWDatabase

Maintenant que KWDatabase peut gere directement les erreur d'encodage par Set|GetEncodingErrorNumber,
on peut simplifier le code existant:
- KWDatabaseTask se contente de memoriser le nombre d'erreur d'encodage dans la base en entree
- les methodes s'dresse diretecment a la base pour generer les messages liees aux erreur d'encodage
- il n'est plus necessaire de memoriser les erreurs d'encodsage dans les evaluations de predicteur

Impacts principaux:
- KWDatabase::AddEncodingErrorMessage: plus besoin de parametre, on prend celui de la base
- KWPredictorEvaluation: supression de la gestion des EncodingErrorNumber
- KWPredictorEvaluation::Evaluate: memorisation dans la base du EncodingErrorNumber
- message emis a partir de la base, et non de la tache
  - KWClassStats::ComputeStats
  - KWDatabaseCheckTask::CheckDatabase
  - KWDatabaseTransferTask::Transfer
  - KWPredictorEvaluator::EvaluatePredictors
  - CCLearningProblemView::BuildCoclustering

marcboulle committed:
Fix missing return value in KWDRNBClassifier::GetUnknownTargetProb

marcboulle committed:
Fix assertions interpretation bugs in debug mode

marcboulle committed:
Make all GetReadPercentage methods const

marcboulle committed:
Extend generation of encoding errors to external tables

marcboulle committed:
Remove unused local variables in sort algorithms
